Teach TSan about Swift Task and Actor model (on top of new task hook machinery)

yln opened the pull request:
I have identified the following conceptual synchronization points at
which task data and computation can cross thread boundaries. We need to
model these in TSan to avoid false positives:

Awaiting an async task (AsyncTask::waitFuture), which has two cases:

The task has completed (AsyncTask::completeFuture). Everything
that happened during task execution "happened before" before the
point where we access its result. We synchronize on the awaited
task.
The task is still executing: the current execution is suspended and
the waiting task is put into the list of "waiters". Once the awaited
task completes, the waiters will be scheduled. In this case, we
synchronize on the waiting task.
Note: there is a similar relationship for task groups which I still have
to investigate. I will follow-up with an additional patch and tests.

Actor job execution (swift::runJobInExecutorContext):
Job scheduling (swift::swift_task_enqueue) precedes/happens before job
execution. Also all job executions (switching actors or suspend/resume)
are serially ordered.

Note: the happens-before edge for schedule->execute isn't strictly
needed in most cases since scheduling calls through to libdispatch's
dispatch_async_f, which we already intercept and model in TSan.
However, I am trying to model Swift Task semantics to increase the
chance of things to continue to work in case the "task backend" is
switched out.

All of this implemented on top of a new task hook machinery.

rdar://74256733

Commit:
Duplicate select Swift Concurrency tests for TSan

I am putting these tests into a new `Sanitizers/tsan` subfolder so we
can keep the same test file name (avoid need for `tsan-` prefix).  This
should increase the chance that changes to the tests make it to the TSan
duplicate.

I will move and rename existing sanitizer tests (including ASan) tests
to keep consistency in a follow-up PR.

Commit:
Enable TSan for copied tests

Commit:
actor_counters.swift: allow specifying of parameters

This is useful during debugging when manually running the test.

Commit:
Add support for calling TSan's acquire()/release()

Commit:
Add Swift Task hook machinery

Commit:
Teach TSan about Swift Task and Actor model

I have identified the following conceptual synchronization points at
which task data and computation can cross thread boundaries.  We need to
model these in TSan to avoid false positives:

Awaiting an async task (`AsyncTask::waitFuture`), which has two cases:
1) The task has completed (`AsyncTask::completeFuture`).  Everything
   that happened during task execution "happened before" before the
   point where we access its result.  We synchronize on the *awaited*
   task.
2) The task is still executing: the current execution is suspended and
   the waiting task is put into the list of "waiters".  Once the awaited
   task completes, the waiters will be scheduled.  In this case, we
   synchronize on the *waiting* task.

Note: there is a similar relationship for task groups which I still have
to investigate.  I will follow-up with an additional patch and tests.

Actor job execution (`swift::runJobInExecutorContext`): Job scheduling
(`swift::swift_task_enqueue`) precedes/happens before job execution.
Also all job executions (switching actors or suspend/resume) are
serially ordered.

Note: the happens-before edge for schedule->execute isn't strictly
needed in most cases since scheduling calls through to libdispatch's
`dispatch_async_f`, which we already intercept and model in TSan.
However, I am trying to model Swift Task semantics to increase the
chance of things to continue to work in case the "task backend" is
switched out.

Radar-Id: rdar://74256733

Commit:
Adopt new hook in place of swift_task_enqueueGlobal_hook

CodaFi commented:
Please make sure you test this patch against the Apple Silicon host. We are currently seeing issues with TSAN there and may have to disable these tests under the banner of rdar://75006869

Commit:
Add positive TSan tests for Swift Concurrency

Add tests that check that TSan detects true positives in simple examples
for both actors and async tasks.

yln commented:
Closing in favor of #36216.
